module/zfs/dsl_scan.c

```diff
@@ -1529,7 +1529,7 @@ dsl_scan_prefetch_thread(void *arg)
 		/* issue the prefetch asynchronously */
 		(void) arc_read(scn->scn_zio_root, scn->scn_dp->dp_spa,
 		    &spic->spic_bp, dsl_scan_prefetch_cb, spic->spic_spc,
-		    ZIO_PRIORITY_ASYNC_READ, zio_flags, &flags, &spic->spic_zb);
+		    ZIO_PRIORITY_SCRUB, zio_flags, &flags, &spic->spic_zb);
```

Comment:
Given the intent of this fix, do we still need to change the priority to SCRUB? Will these inherit the priority of a demand read?

Reply:
I think these IOs logically should be rate-limited based on the scrub vdev queue. Incidentally, this change resolves another small issue where currently `zpool iostat -q` scrub IOs are listed as async reads which is confusing because "I have a scrub running, why isn't it doing anything?"

So its probably not necessary to resolve #6921 but I think it is appropriate.

Reply:
These aren't the real scrub I/Os but just the ones required to traverse the metadata. So the question is should they show up as scrub I/Os or just normal async reads. Did I miss where these get promoted if a demand read is issued on for the same block?

Reply:
> So the question is should they show up as scrub I/Os or just normal async reads.

In the past, these were always given a priority of "async read". It doesn't really make sense to upgrade any of these reads since we don't want them to interfere with normal user IO. The async read category also doesnt seem great now because then they can get intermixed with user prefetches, which isn't the end of the world with the new priority upgrade code. However, it will mean that if the dmu tries to issue a prefetch for a block that could get lost in the sea of scrub prefetches. Therefore, the only priority that makes sense to me is "scrub". 

> Did I miss where these get promoted if a demand read is issued on for the same block?

If you are talking about a user's demand read that will be done by the user's `arc_read()`. If you are talking about the scrub's demand read it doesn't happen with the code the way it is right now for the reasons above.

Reply:
What about the case where we have a demand read for a block that is being prefetched by the scrub metadata traversal? If these were async reads, then it would promote them to sync reads since someone was waiting on it. If the metadata traversal is a scrub priority then the user that issued the demand read must still wait. It feels like the promotion should take place here too.

Reply:
The promotion will take place in this case. The scrub prectch will come in ``arc_read(PRIORITY_SCRUB)` as demand read will come in with `arc_read(PRIORITY_SYNC_READ)`, which will request to upgrade the priority to `PRIORITY_SYNC_READ`

Reply:
Maybe I'm missing something. The promotion happens when we have the following condition:
```if ((hdr->b_flags & ARC_FLAG_PRIO_ASYNC_READ) &&
  			    priority == ZIO_PRIORITY_SYNC_READ) {
```
In the case I was describing, would the PRIORITY_SCRUB I/O set the ARC_FLAG_PRIO_ASYNC_READ flag. I don't think I saw that. If not, then it should so that we promote the I/O. Sorry if I missed that.

Reply:
Oh. I see what you mean. I will fix that.